File: assign3/ruby/lib/rod_cutting.rb
def recursive_rod(prices, rod_length)
  return 0 if prices.empty? || rod_length.zero?

  (1..rod_length).map { |cut| (prices.fetch(cut, 0) + recursive_rod(prices, rod_length - cut)) }.max
end

def memoized_rod(prices, rod_length)
  compute_max_prices = ->(max_prices, length) {
    return max_prices.merge({ length => 0 }) if prices.empty? || length.zero?

    return max_prices if max_prices[length]

    (1..length).reduce(max_prices) { |accumulated_max_prices, cut|
      (max_prices_computed = compute_max_prices.call(accumulated_max_prices, (length - cut))
      max_prices_for_part = prices.fetch(cut, 0) + max_prices_computed.fetch(length - cut, 0)

      max_prices_computed.merge({ length => [max_prices_for_part, max_prices_computed.fetch(length, 0)].max}))
    }
  }
  compute_max_prices.call({}, rod_length)[rod_length]
end

File: assign3/scala/src/main/scala/RodCutting.scala
object RodCutting {

  def recursiveRod(prices: Map[Int, Int], rodLength: Int): Int = {
    if (prices.isEmpty || rodLength == 0) {
      return 0
    }
    
    (1 to rodLength).map(cut => prices.getOrElse(cut, 0) + recursiveRod(prices, rodLength - cut)).max
  }

  def memoizedRod(prices: Map[Int, Int], rodLength: Int): Int = {
    def computeMaxPrices(maxPrices: Map[Int, Int], rodLength: Int): Map[Int, Int] = {
      if (maxPrices.contains(rodLength)) {
        return maxPrices
      }

      if (rodLength == 0 || prices.isEmpty) {
        return maxPrices + (rodLength -> 0)
      }

      (1 to rodLength).foldLeft(maxPrices) { (accumulatedMaxPrices, cut) =>
        val maxPricesComputed = computeMaxPrices(accumulatedMaxPrices, rodLength - cut)

        val maxPricesForPart = prices.getOrElse(cut, 0) + maxPricesComputed.getOrElse(rodLength - cut, 0)

        maxPricesComputed + (rodLength -> math.max(maxPricesForPart, maxPricesComputed.getOrElse(rodLength, 0)))
      }
    }

    computeMaxPrices(Map(), rodLength)(rodLength)
  }
}
